src/Chainweb/Miner/POW.hs

```diff
@@ -132,120 +156,245 @@ powMiner logFun conf nid cutDb = runForever logFun "POW Miner" $ do
-                        --         total
-                        --         (take 30 targetBits)
-                        --         (drop 226 targetBits)
-                        --         (pheight newBh)
-                        --         (pweight newBh)
-                        --         (take 8 . drop 5 . show $ _blockHash p)
-                        --         (take 8 . drop 5 . show $ _blockHash newBh)
-                        --         (int (time newBh - time p) / 1000000 :: Float)
-                        --     hFlush stdout
-
-                        pure $! T3 newBh newCut adjustments'
-        loop nonce
+    blockDb :: ChainId -> Maybe BlockHeaderDb
+    blockDb cid = wcdb ^? webBlockHeaderDb . ix cid
+
+    validatePayload :: BlockHeader -> PayloadWithOutputs -> IO ()
+    validatePayload h o = void $ _pactValidateBlock pact h $ toPayloadData o
 
     getTarget
         :: ChainId
         -> BlockHeader
         -> Adjustments
         -> IO (T2 HashTarget Adjustments)
-    getTarget cid bh adjustments = case HM.lookup (_blockHash bh) adjustments of
+    getTarget cid bh as = case HM.lookup (_blockHash bh) as of
         Just (T2 _ t) -> pure $! T2 t adjustments
         Nothing -> case blockDb cid of
             Nothing -> pure $! T2 (_blockTarget bh) adjustments
             Just db -> do
                 t <- hashTarget db bh
                 pure $! T2 t (HM.insert (_blockHash bh) (T2 (_blockHeight bh) t) adjustments)
 
-    blockDb :: ChainId -> Maybe BlockHeaderDb
-    blockDb cid = wcdb ^? webBlockHeaderDb . ix cid
+    toPayloadData d = PayloadData
+              { _payloadDataTransactions = fst <$> _payloadWithOutputsTransactions d
+              , _payloadDataMiner = _payloadWithOutputsMiner d
+              , _payloadDataPayloadHash = _payloadWithOutputsPayloadHash d
+              , _payloadDataTransactionsHash = _payloadWithOutputsTransactionsHash d
+              , _payloadDataOutputsHash = _payloadWithOutputsOutputsHash d
+              }
+
+-- -------------------------------------------------------------------------- --
+--
+
+getAdjacentParents
+    :: (IxedGet s, IxValue s ~ BlockHeader, Index s ~ ChainId)
+    => s
+    -> BlockHeader
+    -> Maybe BlockHashRecord
+getAdjacentParents c p = BlockHashRecord <$> newAdjHashes
+  where
+    -- | Try to get all adjacent hashes dependencies.
+    --
+    newAdjHashes :: Maybe (HM.HashMap ChainId BlockHash)
+    newAdjHashes = iforM (_getBlockHashRecord $ _blockAdjacentHashes p) $ \xcid _ ->
+        c ^?! ixg xcid . to (tryAdj (_blockHeight p))
+
+    tryAdj :: BlockHeight -> BlockHeader -> Maybe BlockHash
+    tryAdj h b
+        | _blockHeight b == h = Just $! _blockHash b
+        | _blockHeight b == h + 1 = Just $! _blockParent b
+        | otherwise = Nothing
+
+-- | Run a mining loop. Updates creation time and nonce.
+-- The result is guaranteed to have a valid nonce.
+--
+mine :: BlockHeader -> Nonce -> IO BlockHeader
+mine h nonce = do
```

Comment:
Unused?

Reply:
Yes. Should we delete it and only keep the fast miner? I was wondering if this function may still be useful for testing and debugging.

Reply:
I think we should just keep the fast miner.